subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/incremental/deps/ClassSetAnalysisData.java

```diff
@@ -131,7 +133,6 @@ public ClassSetAnalysisData reduceToTypesAffecting(ClassSetAnalysisData other) {
                 usedClasses.add(cls);
             }
         }
-        usedClasses.addAll(classesToConstants.keySet());
```

Comment:
This was unnecessarily inefficient - We were adding the constants before the transitive hull operation. But constants are constant and can't be affected by others, so we didn't need all their upstream dependencies. This is now fixed, along with using less constant holders when possible. This had by far the biggest impact on the analysis file.